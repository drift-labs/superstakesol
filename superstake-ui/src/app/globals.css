@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
	/* --app-gradient: theme('backgroundImage.light-app-gradient');
	--app-gradient-rotate: theme('backgroundImage.light-app-gradient-rotate'); */
	--body-bg: theme('colors.neutrals.90');

	--container-bg: theme('colors.neutrals.100');
	--container-bg-selected: theme('colors.neutrals.90');
	--container-border: theme('colors.black');

	--tooltip-bg: theme('colors.neutrals.100');

	--accent-pink: theme('colors.pink.80');

	--button-border: theme('colors.black');

	--tab-bg: theme('colors.white');
	--tab-bg-selected: theme('colors.pink.90');
	--tab-bg-success: theme('colors.green.80');

	--text-main: theme('colors.neutrals.0');
	--text-default: theme('colors.neutrals.0');
	--text-header: theme('colors.neutrals.10');
	--text-label: theme('colors.neutrals.20');
	--text-disabled: theme('colors.neutrals.70');
	--text-positive-green: theme('colors.green.50');
	--text-negative-red: theme('colors.red.50');
	--text-link: theme('colors.pink.60');

	/* For rpc status etc. */
	--status-neutral: theme('colors.neutrals.70');
	--status-positive: theme('colors.green.30');
	--status-negative: theme('colors.red.50');
}

* {
	box-sizing: border-box;
}

/* Gradients */

.gradient-body-bg {
	background: radial-gradient(
			circle at 10% -10%,
			rgba(255, 224, 206, 1) 20%,
			rgba(255, 224, 206, 0) 60%
		),
		radial-gradient(
			circle at 100% 80%,
			rgba(255, 224, 206, 1) 0%,
			rgba(255, 224, 206, 0) 30%
		),
		radial-gradient(
			circle at 90% -20%,
			rgba(255, 82, 165, 1) 20%,
			rgba(255, 82, 165, 0) 70%
		),
		radial-gradient(
			circle at 0% 100%,
			rgba(235, 111, 255, 1) 30%,
			rgb(235, 111, 255, 0) 80%
		),
		#fafafa;
}

.gradient-border-tooltip {
	position: relative;
}

.gradient-border-tooltip::before {
	content: '';
	position: absolute;
	width: 100%;
	height: 100%;
	box-sizing: content-box;
	padding: 2px;
	top: -2px;
	left: -2px;
	background: linear-gradient(
		90deg,
		rgba(255, 0, 117, 1) -20%,
		rgba(255, 201, 172, 1) 50%,
		rgba(255, 0, 117, 1) 120%
	);
}

.gradient-slider-bg,
.rc-slider-track {
	background: linear-gradient(90deg, #ffe1d1 0%, #ff73b3 153.91%);
}

.skeleton-value-placeholder-animate {
	background: linear-gradient(
		90deg,
		#fafafa 0%,
		#eee 33.33%,
		#fafafa 66.66%,
		#eee 100%
	);
	animation: skeleton 1s infinite linear reverse;
	background-size: 300%;
	background-position: 0% 0%;
}

@keyframes skeleton {
	0% {
		background-position: 0% 0%;
	}
	100% {
		background-position: 100% 0%;
	}
}

input {
	outline: none;
}

input,
textarea,
button,
select,
a {
	-webkit-tap-highlight-color: transparent;
}

/* This has to be done as a class because I don't think tailwind lets you do a media query and first-child / last-child together with just the built-in classnames */
@media (min-width: 1024px) {
	.vault-content-panel:first-child {
		border-top-right-radius: 0;
		border-bottom-right-radius: 0;
	}
	.vault-content-panel:last-child {
		border-top-left-radius: 0;
		border-bottom-left-radius: 0;
		border-left: 0;
	}
}

/* Override rc-slider classes */
.rc-slider-custom .rc-slider-rail {
	height: 16px;
	border: 2px solid var(--container-border);
	background-color: var(--container-bg);
	border-radius: 8px;
}
.rc-slider-custom .rc-slider-track {
	height: 16px;
	border: 2px solid var(--container-border);
	border-radius: 8px;
}
.rc-slider-custom .rc-slider-handle,
.rc-slider-custom .rc-slider-handle:hover,
.rc-slider-custom .rc-slider-handle-dragging {
	width: 28px;
	height: 30px;
	border: var(--container-border) 2px solid !important;
	border-radius: 16px;
	box-shadow: inset 0 -2px 0 var(--container-border) !important;
	outline: none !important;
	background-color: var(--container-bg);
	opacity: 1;
	margin-top: -8px;
}

.rc-slider-disabled {
	background-color: transparent !important;
}

.rc-slider-disabled .rc-slider-track {
	opacity: 40% !important;
}
