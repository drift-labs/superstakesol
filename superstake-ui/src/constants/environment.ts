import { DriftEnv } from '@drift-labs/sdk';
import {
	EnvironmentConstants,
	Initialize as InitializeCommon,
} from '@drift/common';

const driftEnv =
	process.env.NEXT_PUBLIC_DRIFT_ENV === 'mainnet-beta'
		? 'mainnet-beta'
		: ('devnet' as DriftEnv);

InitializeCommon(driftEnv);

type EnvironmentVariables = {
	driftEnv: DriftEnv;
	nextEnv: string | undefined;
	isDev: boolean | undefined;
	basePollingRateMs: number;
	rpcOverride: string | undefined;
	historyServerUrl: string;
};

const Env: EnvironmentVariables = {
	driftEnv,
	nextEnv: process.env.NEXT_PUBLIC_ENV,
	isDev:
		!process.env.NEXT_PUBLIC_ENV ||
		['local', 'master', 'devnet'].includes(process.env.NEXT_PUBLIC_ENV),
	basePollingRateMs: process.env.NEXT_PUBLIC_BASE_POLLING_RATE_MS
		? Number(process.env.NEXT_PUBLIC_BASE_POLLING_RATE_MS)
		: 1000,
	rpcOverride: process.env.NEXT_PUBLIC_RPC_OVERRIDE,
	historyServerUrl:
		process.env.NEXT_PUBLIC_DRIFT_ENV === 'mainnet-beta'
			? EnvironmentConstants.historyServerUrl.mainnet
			: EnvironmentConstants.historyServerUrl.dev,
};

export default Env;
