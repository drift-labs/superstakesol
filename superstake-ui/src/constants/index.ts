import { EnvironmentConstants } from '@drift/common';

export const MAINNET_RPCS = EnvironmentConstants.rpcs.mainnet;
export const DEVNET_RPCS = EnvironmentConstants.rpcs.dev;

export const SLIPPAGE_TOLERANCE_DEFAULT = 0.5;

export const DEFAULT_METRICS_POLLING_RATE_MS = 10_000;
