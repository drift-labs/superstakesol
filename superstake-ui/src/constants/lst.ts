// LST (Liquid Staking Tokens)

import { MainnetSpotMarkets, SpotMarketConfig } from '@drift-labs/sdk';

export type LST = {
	symbol: string; // symbol is added as a config because JitoSOL's symbol in the spotMarket is jitoSOL
	driftAccountName: string;
	spotMarket: SpotMarketConfig;
	logoUrl: string;
	// maxLeverage set manually for now, but would be nice if we make it derived from the asset weight later on
	maxLeverage: number;
	// Default leverage to start the form out with when switching to the lst
	defaultLeverage: number;
};

export const M_SOL: LST = {
	symbol: 'mSOL',
	driftAccountName: 'Super Stake SOL',
	spotMarket: MainnetSpotMarkets[2],
	logoUrl: '/mSol.svg',
	maxLeverage: 3,
	defaultLeverage: 2
};

export const JITO_SOL: LST = {
	symbol: 'JitoSOL',
	driftAccountName: 'Super Stake JitoSOL',
	spotMarket: MainnetSpotMarkets[6],
	logoUrl: '/jitoSol.svg',
	maxLeverage: 1.8,
	defaultLeverage: 1.5
};

// export const B_SOL: LST = {
// 	symbol: 'bSOL',
// 	driftAccountName: 'Super Stake bSOL',
// 	spotMarket: MainnetSpotMarkets[8],
// 	logoUrl: '/bSol.svg',
// 	maxLeverage: 1.9,
// 	defaultLeverage: 1.5
// };

export const ALL_LST: LST[] = [M_SOL, JITO_SOL];
