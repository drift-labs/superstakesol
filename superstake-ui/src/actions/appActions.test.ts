
import {describe, expect, test} from '@jest/globals';
import createAppActions from './appActions';
import { JITO_SOL, M_SOL } from '../constants/lst';

// Todo how do we make a good mock store that can hold state but also be snapshotted?
// This is going to take some thought.
// 
// const storeMock = () => {
//     let state = {};
//     return {
//         get: (key) => state[key],
//         set: (key, value) => {
//         state[key] = value;
//         },
//     };
// }

const mockGet = jest.fn();
const mockSet = jest.fn();
const mockGetCommon = jest.fn();
const mockSetCommon = jest.fn();

describe('src/actions/appActions.ts', () => {
  let actions;

  afterEach(() => {
    mockGet.mockReset();
    mockSet.mockReset();
    mockGetCommon.mockReset();
    mockSetCommon.mockReset();
  });

  test('creates app actions without error', () => {
    actions = createAppActions(mockGet, mockSet, mockGetCommon, mockSetCommon);

    expect(actions).toBeDefined();

    expect(actions).toMatchSnapshot();
  });

  describe('switchActiveLst action', () => {
    test('switchActiveLst sets new state with valid LSTs', () => {
        const jitoSolSymbol = JITO_SOL.symbol;
        const mSolSymbol = M_SOL.symbol;

        actions.switchActiveLst(jitoSolSymbol);
        const setter = mockSet.mock.calls[0][0];
        const state = {};
        setter(state);
        expect(state).toMatchSnapshot();

        actions.switchActiveLst(mSolSymbol);
        const setter2 = mockSet.mock.calls[0][1];
        setter(state);
        expect(state).toMatchSnapshot();
    });

    test('switchActiveLst throws an error with invalid LST', () => {
        const fakeSymbol = 'longLiveSolana';

        expect(() => {
            actions.switchActiveLst(fakeSymbol);
        }).toThrowErrorMatchingSnapshot();
    });
  })
});
