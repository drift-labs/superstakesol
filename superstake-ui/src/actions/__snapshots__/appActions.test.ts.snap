// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/actions/appActions.ts creates app actions without error 1`] = `
{
  "handleSuperStakeDeposit": [Function],
  "handleSuperStakeRepayBorrow": [Function],
  "handleSuperStakeWithdrawal": [Function],
  "resetCurrentUserData": [Function],
  "switchActiveLst": [Function],
  "switchActiveSubaccount": [Function],
  "switchSubaccountToActiveLst": [Function],
  "updateCurrentUserData": [Function],
}
`;

exports[`src/actions/appActions.ts switchActiveLst action switchActiveLst sets new state with valid LSTs 1`] = `
{
  "activeLst": {
    "defaultLeverage": 1.4,
    "driftAccountName": "Super Stake JitoSOL",
    "logoUrl": "/jitoSol.svg",
    "maxLeverage": 1.8,
    "spotMarket": {
      "marketIndex": 6,
      "mint": "J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn",
      "oracle": "7yyaeuJ1GGtVBLT2z2xub5ZWYKaNhF28mj1RdV4VDFVk",
      "oracleSource": {
        "pyth": {},
      },
      "precision": "3b9aca00",
      "precisionExp": "09",
      "serumMarket": "DkbVbMhFxswS32xnn1K2UY4aoBugXooBTxdzkWWDWRkH",
      "symbol": "jitoSOL",
    },
    "symbol": "JitoSOL",
  },
  "stakeUnstakeForm": {
    "amountToStakeString": "",
    "amountToStakeUncappedString": "",
    "amountToUnstakeString": "",
    "isMaxStake": false,
    "isMaxUnstake": false,
    "leverageToUse": 1.4,
    "unstakeLeverage": 0,
  },
}
`;

exports[`src/actions/appActions.ts switchActiveLst action switchActiveLst sets new state with valid LSTs 2`] = `
{
  "activeLst": {
    "defaultLeverage": 1.4,
    "driftAccountName": "Super Stake JitoSOL",
    "logoUrl": "/jitoSol.svg",
    "maxLeverage": 1.8,
    "spotMarket": {
      "marketIndex": 6,
      "mint": "J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn",
      "oracle": "7yyaeuJ1GGtVBLT2z2xub5ZWYKaNhF28mj1RdV4VDFVk",
      "oracleSource": {
        "pyth": {},
      },
      "precision": "3b9aca00",
      "precisionExp": "09",
      "serumMarket": "DkbVbMhFxswS32xnn1K2UY4aoBugXooBTxdzkWWDWRkH",
      "symbol": "jitoSOL",
    },
    "symbol": "JitoSOL",
  },
  "stakeUnstakeForm": {
    "amountToStakeString": "",
    "amountToStakeUncappedString": "",
    "amountToUnstakeString": "",
    "isMaxStake": false,
    "isMaxUnstake": false,
    "leverageToUse": 1.4,
    "unstakeLeverage": 0,
  },
}
`;

exports[`src/actions/appActions.ts switchActiveLst action switchActiveLst throws an error with invalid LST 1`] = `""No LST exists with symbol longLiveSolana"`;
